apps/labs/posts/jupyter-everywhere.md

```diff
@@ -0,0 +1,234 @@
+- The lack of a database to store user data and notebooks grouped by accounts, such as by OAuth providers like Google or Microsoft. This limitation can be circumvented by using alternative methods for notebook sharing and distribution, which we discuss later in this post.
+
+## The story of Jupyter Everywhere
+
+Here, we start by describing our journey building the application from the ground up – from a no-frills prototype to a full-fledged JupyterLite extension.
+
+<!-- architecture diagram here -->
+
+### Customising JupyterLite, and key features of Jupyter Everywhere
+
+Jupyter Everywhere, as we mentioned earlier, is built on top of JupyterLite. JupyterLite is a distribution of JupyterLab that replaces Jupyter's server-side components with standards for in-browser communication with language kernels either in JavaScript or compiled to WebAssembly (WASM), shims for server-side Jupyter APIs, and in-browser file systems for storing user content and settings.
+
+Initially, we started developing the application as a JupyterLab extension to prototype the functionality we wanted to build first: interacting with the file system and downloading notebooks in various formats. This enabled quick creation of the proof of concept, without the need to recompile JupyterLite during the initial iterations. Once we had a clearer idea of the UI and user workflows that needed to be implemented, we switched to a JupyterLite application.
+
+### Facilitating notebook sharing and distribution
+
+Jupyter Everywhere aims to simplify the sharing and distribution of notebooks among students and educators. Given the lack of a backend database to store user data and notebooks grouped by accounts, we had to devise alternative methods for notebook sharing. To this end, our collaborators at CourseKata developed a sharing service that supports authentication via JWT (JSON Web Tokens). This sharing service allows authentication through various API endpoints to authenticate and refresh tokens, and to upload notebooks to a server. The server is connected to the Jupyter Everywhere frontend via a REST API. The notebook contents are stored in an AWS S3 bucket, with appropriate security measures to ensure that only authenticated users can upload and access notebooks. The database is a managed PostgreSQL instance that stores notebooks as `.ipynb` files, with sharing-related metadata embedded as JSON fields in the notebook's "metadata" field.
+
+The sharing service provides two key features:
+- Uploading notebooks to a server for storage and sharing
+- Generating sharing IDs based on a hash of the current session and the notebook state, and readable IDs as a mnemonic for sharing purposes via a database of aquatic animal names and adjectives to make it more fun and engaging for students.
+
+<!-- Add diagram from Adam's architecture-overview.md on Slack -->
```

Comment:
Similar to the section above, maybe this architecture diagram should go before the paragraph that discusses it?